@cothority/byzcoin meaningful instruction print

nkcr opened the pull request:
**What this PR does**

This PR implements a `instruction.getArranged()` function on @cothority/byzcoin that returns a general-purpose JSON representation of an instruction. This representation is contract aware in the sense that arguments should have a meaningful string representation, based on each different contract.

The JSON has the following structure:

```
     *  {
     *      status: 0 | 1,   // Non-zero means an error occured
     *      type: "spawn|invoke|delete|UNKNOWN",
     *      contract: "config|darc|...",
     *      args: [
     *          {
     *              "name": "<NAME>",
     *              "value": "<VALUE>",    // Arg value or its summary
     *              "full": "<FULL VALUE>" // Optional. useful if the argument
     *                                     // have a very long representation.
     *          }
     *          ...
     *      ]
     *  }
```

This functionality is much needed for the Columbus project (blockchain browser). I suggest we include that as an experimental feature until the end of that project (January 2021), where at that point we will have a definitive structure.

Feel free to discuss, I'm open to discussion.

@ineiti's requests:
- [ ] Use an abstract class that the Beautifiers have to implement. Perhaps something like this would work: https://stackoverflow.com/questions/43723313/typescript-abstract-class-static-method-not-enforced
- [x] Give longer names to the classes, this will also avoid any name-clashing:
    - `Config` -> `ConfigBeautify`
    - `Darc` -> `DarcBeautify`
- [x] The tests should pass.
- [x] Squash the commits to have 1 or 2 commits.
- [x] add a `beautifier/index.ts` and import only this. That way you're free to re-arrange any files in there and the user can treat each sub-directory as a package without having to care about the files

ineiti commented on external/js/cothority/src/byzcoin/client-transaction.ts:
```diff
@@ -364,6 +366,73 @@ export class Instruction extends Message<Instruction> {
+     *  }
+     * 
+     */
+    getArranged(): { [k: string]: any } {
```
The return type should be something more restrictive. From a quick look, I see that all returns contain the `contract` and `args` field.

ineiti commented on external/js/cothority/src/byzcoin/arrange/config.ts:
```diff
@@ -0,0 +1,57 @@
+import {Argument} from "../index";
+import {Darc as d} from "../../darc";
```
I would not shorten the imports here - it's perhaps a little bit faster to write, but so much harder to read.

nkcr replied:
I agree. Just followed an ongoing practice in the actual code though.

ineiti commented on external/js/cothority/src/byzcoin/arrange/utils.ts:
```diff
@@ -0,0 +1,32 @@
+ */
+export class Utils {
+
+    // transform a protobuf encoded number to a number
```
I don't think it's a protobuf, but rather a `VarInt`. Also, I would directly import the library from here:

https://github.com/chrisdickinson/varint/blob/master/decode.js

Which keeps the range check that is removed in this copy/paste...

ineiti commented on external/js/cothority/src/byzcoin/client-transaction.ts:
```diff
@@ -364,6 +366,73 @@ export class Instruction extends Message<Instruction> {
+            case Instruction.typeDelete:
+                res.type = "delete";
+                break;
+        }
```
Missing `default` case: there should be either a "unknown class", or even an error thrown. Else somebody using the method might be surprised that it doesn't work...

ineiti replied:
Now it's cleaned up, so it's OK. But in general we should try to break code culture only when it goes into 'best practices'. And using `status` instead of `throw new Error` is not a best practice.

ineiti commented on external/js/cothority/src/byzcoin/config.ts:
```diff
@@ -46,6 +46,19 @@ export default class ChainConfig extends Message<ChainConfig> {
+    /**
+     * Return a string representation of a config
+     */
+    toString(): string {
```
Couldn't that use directly the new pretty-printing?

nkcr replied:
I would rather not. pretty-printing is about printing a transaction, toString() is about printing an instance.

ineiti replied:
I suppose you'll need to print instances, too. At least that's what I'm doing in my `ol-explorer`, and it's handy - I had to undo a deletion, again ;)

nkcr replied:
Sure, something we can still add

ineiti commented:
Very nice. The following parts need to be improved, however:
- programming
  - use interfaces for the classes
  - use interfaces for the data structures
  - add `default` clauses in the `case`s
  - find a better name than `arrange` - pretty-printer, toJSON, beautifier, printDetail, toHTML, ...
- functionality
  - add interpretation of the instance itself
  - add link suggestions to the structures

If it's not to be included anytime soon, you can `Convert to draft`.

nkcr commented:
Thank you for the review. I addressed most of the comments:

> use interfaces for the classes

I can't, as all the methods are static

> functionality: ...

I'd like not to implement too much functionality at first, if that's ok I'd like to include this first iteration and add more on top of that when the use cases come along the way.

ineiti commented on external/js/cothority/src/byzcoin/beautifier/config.ts:
```diff
@@ -0,0 +1,59 @@
+import { ChainConfig } from "../index";
+import { IArgsSchema } from "./utils";
+// tslint:disable-next-line
+const varint = require("varint");
```
I never know - will any of the following work?
```
import * as varint from "varint";
import varint from "varint";
```
???

nkcr replied:
> I never know - will any of the following work?
unfortunately no : /

ineiti commented on external/js/cothority/src/byzcoin/beautifier/config.ts:
```diff
@@ -0,0 +1,59 @@
+ */
+export class Config {
+    static Spawn(args: Argument[]): IArgsSchema {
+        const res = Array<{name: string, value: string, full?: string}>();
```
The `{name...full}` should be an interface, as it's used in many parts.

ineiti commented on external/js/cothority/src/byzcoin/beautifier/utils.ts:
```diff
@@ -0,0 +1,11 @@
+}
+
+export interface IArgsSchema {
+    [index: number]: {name: string, value: string, full?: string};
```
What's the advantage of this against
```
export interface IArgument {
  name: string;
  value: string;
  full?: string;
}
```
And then use it like:
```
static Invoke(args: Argument[]): IArgument[]
```
?

ineiti replied:
Now that I read my own comment I think that `Argument` and `IArgument` are too close. Would `IArgumentDetail` or something be better?

ineiti commented on external/js/cothority/src/byzcoin/client-transaction.ts:
```diff
@@ -320,6 +323,88 @@ export class Instruction extends Message<Instruction> {
+     * The JSON has the following structure:
+     *
+     *  {
+     *      status: 0 | 1
```
What is the status here? Shouldn't it just throw an error in case of error?

nkcr replied:
For the moment non-zero means there is an error, and we only use 1 but we might add additional error code.
I find it more convenient to use than throwing an error.

ineiti replied:
Hmm - thinking out loud (or whatever applies when writing text):

Is it an error if `beautify` cannot return a nice representation?

Perhaps one solution would be to include a default representation which only sets the type-string and iterates over all arguments and sets the `value` to `key:value.toString('hex')`. Then I think the status would make sense: 
- `0` means only the default representation is returned
- `1` means it has been interpreted specifically
- `2` means it also has links (or whatever)

ineiti commented on external/js/cothority/src/byzcoin/client-transaction.ts:
```diff
@@ -320,6 +323,88 @@ export class Instruction extends Message<Instruction> {
+     *
+     *  {
+     *      status: 0 | 1
+     *      type: "spawn|invoke|delete|UNKNOWN",
```
Why `UNKNOWN`? Is this even a possibility?

nkcr replied:
In theory it should never happen

ineiti replied:
Then remove it. If the `Instruction` is neither of the three, then it makes sense to `throw new Error`, as this is something that should not happen.

ineiti reviewed (changes requested):
I put the list of things I think are still needed in the PR message, so you can simply tick the ones you did. Of course feel free to argument any or all of the points ;)

nkcr commented:
@ineiti tests are passing, comments addressed. The proposed trick to use an interface doesn't work.
The reason why Calypso tests are not passing is still mysterious to me, but I don't have time to investigate more. I made it pass by not using `ChainConfig` in `config.ts`.

ineiti reviewed (changes requested):
The switches in the DarcBeautifier and ConfigBeautifier are missing a `default` entry.

nkcr commented:
@ineiti defaults added.

ineiti commented on external/js/cothority/src/byzcoin/beautifier/utils.ts:
```diff
@@ -0,0 +1,14 @@
+
+export interface IBeautifierSchema {
+    status: 0 | 1;
+    type: "spawn" | "invoke" | "delete" | "UNKNOWN";
```
```suggestion
    type: "spawn" | "invoke" | "delete";
```

ineiti commented:
Just remove the `UNKNOWN`, and I think we're good.

ineiti commented:
If you think this is ready, you can move it to R4M. If you think you'll modify it a lot during development of Columbus (which I suppose will happen), you can also teach the student to use a branch of cothority for his development.
